contracts/utils/structs/EnumerableSet.sol

```diff
@@ -114,6 +114,18 @@ library EnumerableSet {
         }
     }
 
+    /**
+     * @dev Removes all the values from a set. O(n).
+     *
+     * WARNING: Developers should keep in mind that this function has an unbounded cost and using it may render the
+     * function uncallable if the set grows to the point where clearing it consumes too much gas to fit in a block.
+     */
+    function _clear(Set storage set) private {
+        for (uint256 i = _length(set); i > 0; --i) {
+            _remove(set, _at(set, i - 1));
+        }
```

Comment:
This feels very wastefull:
- for each iteration, there is a `lastIndex` lookup that identifies that the last element is in fact being removed (1 unecessary sload)
- for each iteration, there is a pop operation that does a sstore for the length with an intermediary value. 
  - that pop also zeros the values, which is honestly not necessary.
  
IMO, this would be much cheaper:
   ```suggestion
        uint256 length = _length(set);
        for (uint256 i = 0; i < length; ++i) {
            delete set._positions[set._values[index]];
        }
        Arrays.unsafeSetLength(set._values, 0);
```

Reply:
Is it not worth it to clear the storage for the gas refund?

Reply:
Also, tx with gas refunds are computed as:
- first we charge the gas for the update (+2900 gas here, because the slot is warm) 
- then at the end we apply a refund (-5000 gas)

The `+2900 * length` part accumulates, and may cause the transaction to get "out of gas", because the refund is not applied immediatly.

So when seeking the refund, you also increase the risk of the tx becoming to expensive to be executable.

Reply:
(There is also a risk that with Verkle the refund may go away)

Reply:
Something else to note is we already have `n` refunds from the mapping so as is, the tx would need to consume a minimum of of `n*5*4800` gas to make use of the refunds. You are probably right that the refund would often go unused if we deleted the array contents as well (requires `n*48000` gas to be used).